beater/v2_handler_test.go

```diff
@@ -0,0 +1,223 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"testing/iotest"
+	"time"
+
+	"github.com/elastic/apm-server/model"
+	errorm "github.com/elastic/apm-server/model/error"
+	"github.com/elastic/apm-server/model/metric"
+	"github.com/elastic/apm-server/model/span"
+	"github.com/elastic/apm-server/model/transaction"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/apm-server/transform"
+)
+
+func validMetadata() string {
+	return `{"metadata": {"service": {"name": "myservice", "agent": {"name": "test", "version": "1.0"}}}}`
+}
+
+func TestV2Handler(t *testing.T) {
+	var transformables []transform.Transformable
+	var reportedTCtx *transform.Context
+	report := func(ctx context.Context, p pendingReq) error {
+		transformables = append(transformables, p.transformables...)
+		reportedTCtx = p.tcontext
+		return nil
+	}
+
+	c := defaultConfig("7.0.0")
+
+	handler := (&v2Route{backendRouteType}).Handler(c, report)
+
+	tx1 := "tx1"
+	timestamp, err := time.Parse(time.RFC3339, "2018-01-01T10:00:00Z")
+	assert.NoError(t, err)
+	reqTimestamp, err := time.Parse(time.RFC3339, "2018-01-02T10:00:00Z")
+	assert.NoError(t, err)
+
+	for idx, test := range []struct {
+		body         string
+		contentType  string
+		err          *StreamResponse
+		expectedCode int
+		reported     []transform.Transformable
```

Comment:
Could you add a test for invalid headers, ERR_INVALID_JSON, ERR_SHUTTING_DOWN, .. please

Reply:
thanks! There's a test for invalid content type headers. Did you have other invalid headers in mind?

Reply:
Added tests for ERR_INVALID_JSON, ERR_SHUTTING_DOWN and found an actual bug ;) thanks!